# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Hardikkk-1209/ML_Pipeline/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane: Refresh / Content Opportunity Scoring

I chose the **Refresh / Content Opportunity Scoring** lane because I want to help a content or SEO team decide which pages should be reviewed first. The project will use observable search and content-performance signals to identify pages that show patterns associated with future performance decline. The final goal is a ranked review queue with a priority score and reason codes, rather than an automatic claim that a page must be refreshed.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

## 2. The question: decision, action, and cost of a wrong call

### Research question

Can historical search-performance signals be used to prioritize content pages that are more likely to experience future performance decline?

### Decision

The work should help a content or SEO team decide **which pages to review for a possible content refresh first**.

### Action

A content editor or SEO analyst can start with the highest-priority pages, inspect the evidence behind their score, and decide whether to refresh the content, review its search presentation, or simply monitor it.

### Cost of a wrong call

A false positive can waste limited editor time on a page that did not need attention. A false negative can cause a genuinely declining page to be missed. Because the output is a ranked queue, we will pay particular attention to whether the highest-ranked pages contain a useful share of future declines.

### Why data and ML can help

The warehouse contains several observable signals, including search visibility, clicks, CTR, position, engagement, content age, and query-level information. A simple rule may miss useful combinations of these signals. We will first build a simple baseline and only keep an ML model if it provides useful improvement on an honest validation split.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [3]:
import pandas as pd
from pathlib import Path
data_path = Path("../../data/raw/content_refresh_anonymized.csv")
df = pd.read_csv(data_path)
print(f"Number of rows: {len(df):,}")
print(f"Number of columns: {len(df.columns):,}")
print(f"Average 90-day impressions: {df['impressions_90d'].mean():,.2f}")
if "trend_direction" in df.columns:
    declining_pct = (df["trend_direction"].eq("down").mean()) * 100
    print(f"Percentage of pages with downward trend: {declining_pct:.2f}%")
print("\nRelevant columns:")
print([
    col for col in [
        "content_id",
        "impressions_90d",
        "clicks_90d",
        "sessions_90d",
        "ctr",
        "avg_position",
        "trend_direction",
        "content_age_days"
    ]
    if col in df.columns
])

Number of rows: 30,000
Number of columns: 44
Average 90-day impressions: 5,200.37
Percentage of pages with downward trend: 54.21%

Relevant columns:
['content_id', 'impressions_90d', 'clicks_90d', 'sessions_90d', 'ctr', 'avg_position', 'trend_direction', 'content_age_days']


### What these numbers suggest

The starter dataset contains **30,000 content records and 44 columns**, with an average of **5,200.37 90-day impressions per page**. About **54.21% of pages have a downward trend** in this starter snapshot. This gives us enough variation to investigate whether observable performance signals can help prioritize pages for review.

These are descriptive numbers only. They do not show that any signal causes a decline, and the starter snapshot is only a starting point for the capstone.

## 4. Careful words: what I can and cannot claim

### What I can claim

I can report patterns observed in the data and test whether current or historical signals are useful for prioritizing pages that later experience a defined performance decline. I can compare a simple baseline with an ML model and report whether the model improves the chosen ranking or classification metric on an honest validation split.

### What I cannot claim

I cannot claim that refreshing a page will cause its traffic or ranking to improve. I cannot claim that the model has discovered Google's ranking algorithm or a causal Google ranking factor. I also cannot present a model score as certainty about what will happen to an individual page.

The final results will therefore use careful language such as **observed, associated with, suggests, directional, and decision-support**.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card.